File: CS/ITypedList/ViewModels/MainViewModel.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Linq;
using System.Runtime.CompilerServices;
using DevExpress.Mvvm.CodeGenerators;
using DevExpress.Mvvm.Native;

namespace ITypedList.ViewModels {
    [GenerateViewModel]
    public partial class MainViewModel {
        [GenerateProperty] private ItemCollection _items;

        static MainViewModel() {
            ItemCollection.CustomFields.Add(new ItemPropertyDescriptor(new CustomField("Name", typeof(string))));
            ItemCollection.CustomFields.Add(
                new ItemPropertyDescriptor(new CustomField("CreatedAt", typeof(DateTime))));
        }

        public MainViewModel() {
            _items = new ItemCollection(Enumerable.Range(0, 10).Select(i => new Item { Id = i }));
            _items.ForEach(x => x["Name"] = $"Item {_items.IndexOf(x)}");
            _items.ForEach(x => x["CreatedAt"] = DateTime.Now.AddDays(_items.IndexOf(x)));
        }

        [GenerateCommand]
        public void AddColumn() {
            var fieldCount = ItemCollection.CustomFields.Count;
            ItemCollection.CustomFields.Add(
                new ItemPropertyDescriptor(new CustomField($"Value {fieldCount}", typeof(int))));
            _items.ForEach(x => x[$"Value {fieldCount}"] = _items.IndexOf(x) * fieldCount);
        }
    }

    [GenerateViewModel]
    public partial class Item : INotifyPropertyChanged {
        private readonly Dictionary<string, object> _customFieldValues = new();

        [GenerateProperty] private int _id;

        public object this[string fieldName] {
            get {
                _customFieldValues.TryGetValue(fieldName, out var value);
                return value;
            }

            set {
                _customFieldValues[fieldName] = value;
                OnPropertyChanged(fieldName);
            }
        }

        public event PropertyChangedEventHandler PropertyChanged;

        protected virtual void OnPropertyChanged([CallerMemberName] string propertyName = null) {
            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
        }
    }
}

File: VB/ITypedList/MainViewModel.vb
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Linq
Imports System.Windows.Input
Imports DevExpress.Mvvm
Imports DevExpress.Mvvm.Native

Namespace ITypedList

    Public Class MainViewModel
        Inherits ViewModelBase

        Public ReadOnly Property Items As ItemCollection

        Shared Sub New()
            ItemCollection.CustomFields.Add(New ItemPropertyDescriptor(New CustomField("Name", GetType(String))))
            ItemCollection.CustomFields.Add(New ItemPropertyDescriptor(New CustomField("CreatedAt", GetType(Date))))
        End Sub

        Public Sub New()
            Items = New ItemCollection(Enumerable.Range(0, 10).[Select](Function(i) New Item With {.Id = i}))
            Items.ForEach(Sub(x) x("Name") = $"Item {Items.IndexOf(x)}")
            Items.ForEach(Sub(x) x("CreatedAt") = Date.Now.AddDays(Items.IndexOf(x)))
            AddColumnCommand = New DelegateCommand(AddressOf AddColumn)
        End Sub

        Public Sub AddColumn()
            Dim fieldCount = ItemCollection.CustomFields.Count
            ItemCollection.CustomFields.Add(New ItemPropertyDescriptor(New CustomField($"Value {fieldCount}", GetType(Integer))))
            Items.ForEach(Sub(x) x($"Value {fieldCount}") = Items.IndexOf(x) * fieldCount)
        End Sub

        Public ReadOnly Property AddColumnCommand As ICommand
    End Class

    Public Class Item
        Inherits BindableBase

        Private ReadOnly _customFieldValues As Dictionary(Of String, Object) = New Dictionary(Of String, Object)()

        Public Property Id As Integer
            Get
                Return GetProperty(Function() Me.Id)
            End Get

            Set(ByVal value As Integer)
                SetProperty(Function() Id, value)
            End Set
        End Property

        Default Public Property Item(ByVal fieldName As String) As Object
            Get
                Dim value As Object = Nothing
                _customFieldValues.TryGetValue(fieldName, value)
                Return value
            End Get

            Set(ByVal value As Object)
                _customFieldValues(fieldName) = value
                RaisePropertyChanged(fieldName)
            End Set
        End Property
    End Class
End Namespace
